README
httpio
======

This module provides an alternative to the standard library's
`http.client` module. In contrast to `http.client`, `httpio`:

 - allows you to send multiple requests right after each other without
   having to read the responses first.

 - supports waiting for 100-continue before sending the request body.

 - raises an exception instead of silently delivering partial data if the
   connection is closed before all data has been received.

 - raises one specific exception (ConnectionClosed) if the connection has
   been closed (while http.client connection may raise any of
   BrokenPipeError, BadStatusLineError, ConnectionAbortedError,
   ConnectionResetError, or simply return '' on read)

 - provides a single `HTTPConnection` class to communicate with the
   server (no separate `HTTPResponse` class).

Of course, these features come for a price:

 - Only HTTP 1.1 connections are supported

 - Extra care has to be taken when using non-idempotent http
   methods. This is because if a connection is terminated earlier than
   expected (e.g. because of the server sending an unsupported reply)
   but responses for multiple requests are pending, `httpio` cannot
   determine which requests have already been processed and which ones
   need to be resend.

       
For detailed usage instructions, consult the docstring of the
`httpio.HTTPConnection` class.


Changes.rst
.. currentmodule:: dugong

Release 3.8.2 (2021-07-04)
==========================

* Improved distinction between the DNS server not being available and a hostname not
  existing.


Release 3.8.1 (2021-01-10)
==========================

* Dugong is now compatible with pytest 6.x and newer.

Release 3.8.0 (2020-11-09)
==========================

* Dugong now works under Windows.

Release 3.7.5 (2019-07-17)
==========================

* Fixed a deadlock when resuming coroutines.
* Made compatible with pytest 4.

Release 3.7.4 (2019-01-19)
==========================

* All traffic can now be dumped for debugging purposes using the
  ``DUGONG_TRACEFILE`` environment variable. Make sure to include ``%d``
  somewhere in the filename - it will be replaced with a unique number
  to distinguish between different connections.

Release 3.7.3 (2018-04-29)
==========================

* Minor bugfixes.

Release 3.7.2 (2018-02-11)
==========================

* Now compatible with py.test 3.3

Release 3.7.1 (2017-06-05)
==========================

* Now compatible with Python 3.6.

Release 3.7 (2016-06-20)
========================

* Dugong now supports server responses that specify just ``Connection:
  close`` instead of providing the response length or using chunked
  encoding.

* Dugong now honors the `~ssl.SSLContext.check_hostname` attribute of
  `~ssl.SSLContext` objects.

Release 3.6 (2016-04-23)
========================

* Dugong now uses semantic versioning. This means that
  backwards-incompatible versions (i.e., versions that change the
  existing API in some way) will be reflected in an increase of the
  major version number, i.e. the next backwards-incompatible version
  will have version 4.0.

* Various minor bugfixes.

Release 3.5 (2015-01-31)
========================

* The `is_temp_network_error` function now knows about more kinds of
  temporary errors, in particular also about those that do not have a
  corresponding Python exception (e.g. no route to host).

Release 3.4 (2014-11-29)
========================

* The :file:`examples/extract_links.py` script is now Python 3.3
  compatible again.

* Dugong now has proper proxy support for plain HTTP
  connections. Thanks to Vincent Bernat for the patch.

* Dugong now raises `DNSUnavailable` or `HostnameNotResolvable`
  exceptions instead of the more generic `socket.gaierror` and
  `socket.herror` exceptions where possible.

Release 3.3 (2014-08-06)
========================

* It was possible for some methods to raise `BrokenPipeError`, or
  `ConnectionResetError` instead of `ConnectionClosed` (especially
  under FreeBSD). This has now been fixed.

* It was possible for methods that need to read data from the server
  to raise `AssertionError` in some circumstances. This has been
  fixed.

Release 3.2 (2014-07-27)
========================

* A `HTTPConnection` instance can now be used as a context manager.

* If a connection is closed unexpectedly while request body data is
  being written to the server (i.e., during a call to
  `HTTPConnection.write` or `HTTPConnection.co_write`), dugong now
  pretends that the body has been sent to the server completely (while
  still raising `ConnectionClosed`).

  This makes it possible to catch the exception and nevertheless call
  `~HTTPConnection.read_response` (or
  `~HTTPConnection.co_read_response`) to try to read an error response
  that the server may have sent during the upload (if no response has
  been received, `ConnectionClosed` will be raised again).

* `is_temp_network_error` now actively tries to distinguish between
  permanent and temporary name resolution problems by attempting to
  resolve a number of test hostnames.

* `HTTPConnection` has a new `~HTTPConnection.timeout`
  attribute. Regular `HTTPConnection` methods (i.e., no coroutines)
  will now raise `ConnectionTimedOut` if no data could be send or
  received for *timeout* seconds.

Release 3.1 (2014-06-28)
========================

* Fixed a problem with some testcases failing with a BrokenPipeError.

* Fixed a bug that, in some cases, resulted in additional ``\0`` bytes
  being appended at the end of the response body, or in an incorrect
  `InvalidResponse` exception being raised.

* When trying to continue reading or writing body data after calling
  `HTTPConnection.disconnect`, dugong now raises `ConnectionClosed`
  instead of `AttributeError`.

Release 3.0, (2014-04-20)
=========================

* Major version bump because of backwards incompatible changes.

* Added `HTTPConnection.read_raw` method.

* The `PollNeeded` class now uses the `!select.POLLIN` and
  `!select.POLLOUT` constants instead of `!select.EPOLLIN` and
  `!select.EPOLLOUT` to signal what kind of I/O needs to be
  performed. This makes dugong compatible with systems lacking epoll
  (e.g. FreeBSD).

* The unit tests now check if the host is reachable before trying to
  run the example scripts. This avoids bogus test errors if
  there is no internet connection or if the remote host is down.
  (issue #7).


Release 2.2 (2014-03-14)
========================

* Unittests requiring the `asyncio` module are now skipped if this
  module is not available.


Release 2.1 (2014-03-11)
========================

* Fixed a problem where data was not sent to the server if the syscall
  was interrupted by a signal.

* It is no longer necessary to read from response body at least once
  even if has zero length.

* `PollNeeded.poll` now uses `select.poll` instead of
  `select.select`. This avoids a "filedescriptor out of range"
  exception that may be raised by `select.select` when the
  filedescriptor exceeds some system-specific value.


Release 2.0 (2014-02-23)
========================

* Renamed module from *httpio* to *dugong*.

* The coroutine based API was completely overhauled.

* Introduced `BodyFollowing` class for use with *body* parameter of
  `~HTTPConnection.send_request` method.

* `~HTTPConnection.send_request` now returns a `HTTPResponse` instance

test_httpio.py
'''
test_httpio.py - Unit tests for httpio.py - run with py.test

Copyright (C) 2013 Nikolaus Rath <Nikolaus@rath.org>

This module may be distributed under the terms of the Python Software Foundation
License Version 2.  The complete license text may be retrieved from
http://hg.python.org/cpython/file/65f2c92ed079/LICENSE.
'''

from httpio import HTTPConnection, BUFSIZE
import httpio
from http.server import BaseHTTPRequestHandler, _quote_html
from io import BytesIO
import http.client
import pytest
import time
import ssl
import re
import os
import threading
import socketserver

# We want to test with a real certificate
SSL_TEST_HOST = 'www.google.com'

@pytest.fixture
def ssl_context(path=None):
    context = ssl.SSLContext(ssl.PROTOCOL_SSLv23)
    context.options |= ssl.OP_NO_SSLv2
    context.verify_mode = ssl.CERT_REQUIRED

    if path is None:
        context.set_default_verify_paths()
    elif os.path.isfile(path):
        context.load_verify_locations(cafile=path)
    else:
        context.load_verify_locations(capath=path)

    return context


class TestHTTPServer(threading.Thread):
    def __init__(self):
        super().__init__()
        self.host = 'localhost'
        self.httpd = socketserver.TCPServer((self.host, 0),
                                            TestRequestHandler)
        self.port = self.httpd.socket.getsockname()[1]

    def run(self):
        self.httpd.serve_forever()

    def shutdown(self):
        self.httpd.shutdown()
        self.httpd.server_close()


@pytest.fixture(scope='module')
def http_server(request):
    httpd = TestHTTPServer()
    httpd.start()
    request.addfinalizer(httpd.shutdown)
    return httpd

@pytest.fixture
def conn(request, http_server):
    conn = HTTPConnection(http_server.host, port=http_server.port)
    request.addfinalizer(conn.close)
    return conn

def check_http_connection(ssl_context):
    '''Skip test if we can't connect to ssl test server'''

    try:
        conn = http.client.HTTPSConnection(SSL_TEST_HOST, context=ssl_context)
        conn.request('GET', '/')
        resp = conn.getresponse()
        assert resp.status == 200
    except:
        pytest.skip('%s not reachable but required for testing' % SSL_TEST_HOST)
    finally:
        conn.close()


def readall(conn):
    '''Read from *conn* until EOF, return number of bytes read'''
    
    read = 0
    while True:
        buf = conn.read(BUFSIZE)
        read += len(buf)
        if not buf:
            return read
        
def test_connect_ssl(ssl_context):
    check_http_connection(ssl_context)

    conn = HTTPConnection(SSL_TEST_HOST, ssl_context=ssl_context)
    conn.send_request('GET', '/')
    (method, url, status, reason, header) = conn.read_response()
    assert status == 200
    assert url == '/'
    readall(conn)
    conn.close()

def test_invalid_ssl(ssl_context):
    check_http_connection(ssl_context)

    # Don't load certificates
    context = ssl.SSLContext(ssl.PROTOCOL_SSLv23)
    context.options |= ssl.OP_NO_SSLv2
    context.verify_mode = ssl.CERT_REQUIRED
    conn = HTTPConnection(SSL_TEST_HOST, ssl_context=context)
    with pytest.raises(ssl.SSLError):
        conn.send_request('GET', '/')
    conn.close()
    

def test_get_pipeline(conn):
    
    def read_response():
        if conn.get_current_response():
            conn.read(BUFSIZE)
        else:
            (method, url, status, reason, header) = conn.read_response()

    interrupted = False
    sleeptime = 0.01
    for doc in ('/send_%d_120-byte_chunks' % x for x in range(30)):
        cofun = conn.send_request('GET', doc, via_cofun=True)
        for _ in cofun:
            read_response()
            interrupted = True

        # We want to be interrupted at least once, so wait a little bit
        # before sending the next request
        time.sleep(sleeptime)
        
        if not interrupted:
            sleeptime *= 2
        
    while conn.response_pending():
        read_response()

    assert interrupted


def test_read_identity(conn):
    conn.send_request('GET', '/send_10_bytes')
    (method, url, status, reason, header) = conn.read_response()
    assert status == 200
    assert url == '/send_10_bytes'
    assert readall(conn) == 10

def test_read_chunked(conn):
    conn.send_request('GET', '/send_3_15-byte_chunks')
    (method, url, status, reason, header) = conn.read_response()
    assert status == 200
    assert readall(conn) == 3*15
    
def test_double_read(conn):
    conn.send_request('GET', '/send_10_bytes')
    (method, url, status, reason, header) = conn.read_response()
    assert status == 200
    assert url == '/send_10_bytes'
    with pytest.raises(httpio.StateError):
        (method, url, status, reason, header) = conn.read_response()
    

def writeall(conn, buf):
    while buf:
        off = conn.write(buf)
        buf = buf[off:]

def test_put(conn):
    conn.send_request('PUT', '/allgood', body=b'a nice body string')
    (method, url, status, reason, header) = conn.read_response()
    assert status == 204

def test_body_separate(conn):
    conn.send_request('PUT', '/allgood', body=50)
    writeall(conn, DUMMY_DATA[:50])    
    (method, url, status, reason, header) = conn.read_response()
    assert status == 204
    
def test_write_toomuch(conn):
    conn.send_request('PUT', '/allgood', body=42)
    with pytest.raises(httpio.ExcessBodyData):
        writeall(conn, DUMMY_DATA[:43])
    
def test_write_toolittle(conn):
    conn.send_request('PUT', '/allgood', body=42)
    writeall(conn, DUMMY_DATA[:24])
    with pytest.raises(httpio.StateError):
        conn.send_request('GET', '/send_5_bytes')
        
def test_write_toolittle2(conn):
    conn.send_request('PUT', '/allgood', body=42)
    writeall(conn, DUMMY_DATA[:24])
    with pytest.raises(httpio.StateError):
        conn.read_response()

def test_write_toolittle3(conn):
    conn.send_request('GET', '/send_10_bytes')
    conn.send_request('PUT', '/allgood', body=42)
    writeall(conn, DUMMY_DATA[:24])
    (method, url, status, reason, header) = conn.read_response()
    assert status == 200
    assert url == '/send_10_bytes'
    assert readall(conn) == 10
    with pytest.raises(httpio.StateError):
        conn.read_response()

def test_co_sendfile(conn):
    fh = BytesIO(DUMMY_DATA)
    conn.send_request('PUT', '/allgood', body=len(DUMMY_DATA)//2)
    for _ in conn.co_sendfile(fh):
        pass
    (method, url, status, reason, header) = conn.read_response()
    assert status == 204
    
def test_co_sendfile2(conn):
    fh = BytesIO(DUMMY_DATA)
    conn.send_request('PUT', '/allgood', body=2*len(DUMMY_DATA)-2)
    for _ in conn.co_sendfile(fh):
        pass
    with pytest.raises(httpio.StateError):
        conn.read_response()
    fh.seek(0)
    for _ in conn.co_sendfile(fh):
        pass
    (method, url, status, reason, header) = conn.read_response()
    assert status == 204

def test_100cont(conn):
    conn.send_request('PUT', '/fail_with_403', body=256, expect100=True)
    (method, url, status, reason, header) = conn.read_response()
    assert status == 403
    readall(conn)
    
    conn.send_request('PUT', '/all_good', body=256, expect100=True)
    (method, url, status, reason, header) = conn.read_response()
    assert status == 100
    writeall(conn, DUMMY_DATA[:256])
    (method, url, status, reason, header) = conn.read_response()
    assert status == 204
    
def test_100cont_2(conn):
    conn.send_request('PUT', '/fail_with_403', body=256, expect100=True)

    with pytest.raises(httpio.StateError):
        conn.send_request('PUT', '/fail_with_403', body=256, expect100=True)

def test_100cont_3(conn):
    conn.send_request('PUT', '/fail_with_403', body=256, expect100=True)

    with pytest.raises(httpio.StateError):
        conn.write(b'barf!')

        
def test_tunnel(http_server):
    conn = HTTPConnection('remote_server', proxy=(http_server.host, http_server.port))
    
    conn.send_request('GET', '/send_10_bytes')
    (method, url, status, reason, header) = conn.read_response()
    assert status == 200
    assert url == '/send_10_bytes'
    assert readall(conn) == 10
    conn.close()


def test_request_via_cofun(conn):
    cofun = conn.send_request('GET', '/send_10_bytes', via_cofun=True)
    with pytest.raises(httpio.StateError):
        conn.read_response()

    for _ in cofun:
        pass
    
    (method, url, status, reason, header) = conn.read_response()
    assert status == 200
    assert readall(conn) == 10

def test_read_toomuch(conn):
    conn.send_request('GET', '/send_10_bytes')
    conn.send_request('GET', '/send_8_bytes')
    (method, url, status, reason, header) = conn.read_response()
    assert status == 200
    assert url == '/send_10_bytes'
    assert readall(conn) == 10
    with pytest.raises(httpio.StateError):
        conn.read(8)

        
def test_read_toolittle(conn):
    conn.send_request('GET', '/send_10_bytes')
    (method, url, status, reason, header) = conn.read_response()
    assert status == 200
    assert url == '/send_10_bytes'
    conn.read(8)
    with pytest.raises(httpio.StateError):
        (method, url, status, reason, header) = conn.read_response()

        
def test_current_response(conn):
    assert conn.get_current_response() is None
    conn.send_request('GET', '/send_10_bytes')
    (method, url, status, reason, header) = conn.read_response()
    assert conn.get_current_response() == (method, url)
    conn.read(5)
    assert conn.get_current_response() == (method, url)
    readall(conn)
    assert conn.get_current_response() is None
    

def test_head(conn):
    conn.send_request('HEAD', '/send_10_bytes')
    (method, url, status, reason, header) = conn.read_response()
    assert status == 200
    assert readall(conn) == 0
    
    conn.send_request('HEAD', '/fail_with_317')
    (method, url, status, reason, header) = conn.read_response()
    assert status == 317
    assert readall(conn) == 0


with open(__file__, 'rb') as fh:
    DUMMY_DATA = fh.read()
    
class TestRequestHandler(BaseHTTPRequestHandler):

    server_version = "TestHTTP"
    protocol_version = 'HTTP/1.1'
    
    def handle_expect_100(self):
        if self.handle_errors():
            return
        else:
            self.send_response_only(100)
            self.end_headers()
            return True
        
    def do_GET(self):
        if self.handle_errors():
            return
        
        hit = re.match(r'^/send_([0-9]+)_bytes', self.path)
        if hit:
            len_ = int(hit.group(1))
            self.send_response(200)
            self.send_header("Content-Type", 'application/octet-stream')
            self.send_header("Content-Length", str(len_))
            self.end_headers()
            self.send_dummy_data(len_)
            return

        hit = re.match(r'^/send_([0-9]+)_([0-9]+)-byte_chunks', self.path)
        if hit:
            count = int(hit.group(1))
            len_ = int(hit.group(2))
            self.send_response(200)
            self.send_header("Content-Type", 'application/octet-stream')
            self.send_header("Transfer-Encoding", 'chunked')
            self.end_headers()
            for i in range(count):
                self.wfile.write(('%x\r\n' % len_).encode('us-ascii')) 
                self.send_dummy_data(len_)
                self.wfile.write(b'\r\n')
            self.wfile.write(b'0\r\n\r\n')
            return
        
        self.send_error(500)
        
    def send_dummy_data(self, len_):
        while len_ > 0:
            self.wfile.write(DUMMY_DATA[:len_])
            len_ -= len(DUMMY_DATA)

    def handle_errors(self):
        hit = re.match(r'^/fail_with_([0-9]+)', self.path)
        if hit:
            self.send_error(int(hit.group(1)))
            return True

        if self.command == 'PUT':
            encoding = self.headers['Content-Encoding']
            if encoding and encoding != 'identity':
                self.send_error(415)
                return True

        return False

            
    def do_PUT(self):
        if self.handle_errors():
            return

        len_ = int(self.headers['Content-Length'])
        self.rfile.read(len_)

        self.send_response(204)
        self.end_headers()
        

    def do_CONNECT(self):
        # Just pretend we're the remote server too
        self.send_response(200)
        self.end_headers()
        self.close_connection = 0
        
    def do_HEAD(self):
        if self.handle_errors():
            return

        hit = re.match(r'^/send_([0-9]+)_bytes', self.path)
        if hit:
            len_ = int(hit.group(1))
            self.send_response(200)
            self.send_header("Content-Type", 'application/octet-stream')
            self.send_header("Content-Length", str(len_))
            self.end_headers()
            return

        # No idea
        self.send_error(500)
            
    def send_error(self, code, message=None):
        # Overwritten to not close connection on errors and provide
        # content-length
        try:
            shortmsg, longmsg = self.responses[code]
        except KeyError:
            shortmsg, longmsg = '???', '???'
        if message is None:
            message = shortmsg
        explain = longmsg
        self.log_error("code %d, message %s", code, message)
        # using _quote_html to prevent Cross Site Scripting attacks (see bug #1100201)
        content = (self.error_message_format % {'code': code, 'message': _quote_html(message),
                                               'explain': explain}).encode('utf-8', 'replace')
        self.send_response(code, message)
        self.send_header("Content-Type", self.error_content_type)
        self.send_header("Content-Length", str(len(content)))
        self.end_headers()
        if self.command != 'HEAD' and code >= 200 and code not in (204, 304):
            self.wfile.write(content)


test/conftest.py
import sys
import os.path
import logging
import gc

if sys.version_info < (3,3):
    raise SystemExit('Python version is %d.%d.%d, but Dugong requires 3.3 or newer'
                     % sys.version_info[:3])
# Enable output checks
pytest_plugins = ('pytest_checklogs')

def pytest_addoption(parser):
    group = parser.getgroup("general")
    group._addoption("--installed", action="store_true", default=False,
                     help="Test the installed package.")

    group = parser.getgroup("terminal reporting")
    group._addoption("--logdebug", action="append", metavar='<module>',
                     help="Activate debugging output from <module> for tests. Use `all` "
                          "to get debug messages from all modules. This option can be "
                          "specified multiple times.")

def pytest_configure(config):
    # If we are running from the source directory, make sure that we load
    # modules from here
    basedir = os.path.abspath(os.path.join(os.path.dirname(__file__), '..'))
    if not config.getoption('installed'):
        if (os.path.exists(os.path.join(basedir, 'setup.py')) and
            os.path.exists(os.path.join(basedir, 'dugong', '__init__.py'))):
            sys.path.insert(0, basedir)

        # Make sure that called processes use the same path
        pp = os.environ.get('PYTHONPATH', None)
        if pp:
            pp = '%s:%s' % (basedir, pp)
        else:
            pp = basedir
        os.environ['PYTHONPATH'] = pp

    # When running from VCS repo, enable all warnings
    if os.path.exists(os.path.join(basedir, 'MANIFEST.in')):
        import warnings
        warnings.resetwarnings()
        warnings.simplefilter('default')

    # Configure logging. We don't set a default handler but rely on
    # the catchlog pytest plugin.
    logdebug = config.getoption('logdebug')
    root_logger = logging.getLogger()
    if logdebug is not None:
        logging.disable(logging.NOTSET)
        if 'all' in logdebug:
            root_logger.setLevel(logging.DEBUG)
        else:
            for module in logdebug:
                logging.getLogger(module).setLevel(logging.DEBUG)
    else:
        root_logger.setLevel(logging.INFO)
        logging.disable(logging.DEBUG)
    logging.captureWarnings(capture=True)

# Run gc.collect() at the end of every test, so that we get ResourceWarnings
# as early as possible.
def pytest_runtest_teardown(item, nextitem):
    gc.collect()


test/test_aio.py
'''
test_aio.py - Unit tests for Dugong

Copyright © 2014 Nikolaus Rath <Nikolaus.org>

This module may be distributed under the terms of the Python Software Foundation
License Version 2.  The complete license text may be retrieved from
http://hg.python.org/cpython/file/65f2c92ed079/LICENSE.
'''


if __name__ == '__main__':
    import sys
    import pytest
    sys.exit(pytest.main([__file__] + sys.argv[1:]))

import socket
import logging
from dugong import PollNeeded, POLLIN

try:
    import asyncio
except ImportError:
    import pytest
    pytestmark = pytest.mark.skipif(True,
                                    reason='asyncio module not available')
else:
    from dugong import AioFuture

log = logging.getLogger(__name__)

def read(sock):
    for i in range(3):
        log.debug('yielding')
        yield PollNeeded(sock.fileno(), POLLIN)
        log.debug('trying to read')
        buf = sock.recv(100)
        assert buf.decode() == 'text-%03d' % i
        log.debug('got: %s', buf)

@asyncio.coroutine
def write(sock):
    for i in range(3):
        log.debug('sleeping')
        yield from asyncio.sleep(1)
        buf = ('text-%03d' % i).encode()
        log.debug('writing %s', buf)
        sock.send(buf)

def test_aio_future():
    loop = asyncio.get_event_loop()
    try:
        (sock1, sock2) = socket.socketpair()

        asyncio.Task(write(sock2))
        read_fut = AioFuture(read(sock1))
        read_fut.add_done_callback(lambda fut: loop.stop())
        loop.call_later(6, loop.stop)
        loop.run_forever()

        assert read_fut.done()

        sock1.close()
        sock2.close()
    finally:
        loop.close()


examples/httpcat.py
#!/usr/bin/env python3
'''
Retrieve a list of URLs and print them to stdout.
'''

import sys
import os.path
from io import TextIOWrapper
import re
from urllib.parse import urlsplit

# We are running from the dugong source directory, append it to module path so
# that we can fallback on it if dugong hasn't been installed yet.
if __name__ == '__main__':
    basedir = os.path.abspath(os.path.join(os.path.dirname(sys.argv[0]), '..'))
else:
    basedir = os.path.abspath(os.path.join(os.path.dirname(__file__), '..'))
if (os.path.exists(os.path.join(basedir, 'setup.py')) and
    os.path.exists(os.path.join(basedir, 'dugong', '__init__.py'))):
    sys.path.append(basedir)

# When running from HG repo, enable all warnings
if os.path.exists(os.path.join(basedir, '.hg')):
    import warnings
    warnings.simplefilter('error')

from dugong import HTTPConnection, BUFFER_SIZE

for arg in sys.argv[1:]:
    url = urlsplit(arg)
    assert url.scheme == 'http'
    path = url.path
    if url.query:
        path += '?' + url.query

    with HTTPConnection(url.hostname, url.port) as conn:
        conn.send_request('GET', path)
        resp = conn.read_response()
        if resp.status != 200:
            raise SystemExit('%d %s' % (resp.status, resp.reason))

        # Determine if we're reading text or binary data, and (in case of text),
        # what character set is being used.
        if 'Content-Type' not in resp.headers:
            type_ = 'application/octet-stream'
        else:
            type_ = resp.headers['Content-Type']

        hit = re.match(r'(.+?)(?:; charset=(.+))?$', type_)
        if not hit:
            raise SystemExit('Unable to parse content-type: %s' % type_)
        if hit.group(2):
            charset = hit.group(2)
        elif hit.group(1).startswith('text/'):
            charset = 'latin1'
        else:
            charset = None # binary data

        if charset:
            instream = TextIOWrapper(conn, encoding=charset)
            outstream = sys.stdout
        else:
            instream = conn
            # Since we're writing bytes rather than text, we need to bypass
            # any encoding.
            outstream = sys.stdout.raw

        while True:
            buf = instream.read(BUFFER_SIZE)
            if not buf:
                break
            outstream.write(buf)


examples/pipeline1.py
#!/usr/bin/env python3

import sys
import os.path
from urllib.parse import urlsplit, urlunsplit

# We are running from the dugong source directory, append it to module path so
# that we can fallback on it if dugong hasn't been installed yet.
if __name__ == '__main__':
    basedir = os.path.abspath(os.path.join(os.path.dirname(sys.argv[0]), '..'))
else:
    basedir = os.path.abspath(os.path.join(os.path.dirname(__file__), '..'))
if (os.path.exists(os.path.join(basedir, 'setup.py')) and
    os.path.exists(os.path.join(basedir, 'dugong', '__init__.py'))):
    sys.path.append(basedir)

# When running from HG repo, enable all warnings
if os.path.exists(os.path.join(basedir, '.hg')):
    import warnings
    warnings.simplefilter('error')

# Assemble path list
hostname = None
path_list = []
for url in sys.argv[1:]:
    o = urlsplit(url)
    if hostname is None:
        hostname = o.hostname
        port = o.port
    elif (hostname, port) != (o.hostname, o.port):
        raise SystemExit('Can only pipeline to one host')
    if o.scheme != 'http':
        raise SystemExit('Can only do http')
    path_list.append(urlunsplit(('', '') + o[2:4] + ('',)))


# Code from here on is included in documentation
# start-example
import asyncio
import atexit
from dugong import HTTPConnection, AioFuture

# Get a MainLoop instance from the asyncio module to switch
# between coroutines (and clean up at program exit)
loop = asyncio.get_event_loop()
atexit.register(loop.close)

with HTTPConnection(hostname, port) as conn:
    # This generator function returns a coroutine that sends
    # all the requests.
    def send_requests():
        for path in path_list:
            yield from conn.co_send_request('GET', path)

    # This generator function returns a coroutine that reads
    # all the responses
    def read_responses():
        bodies = []
        for path in path_list:
            resp = yield from conn.co_read_response()
            assert resp.status == 200
            buf = yield from conn.co_readall()
            bodies.append(buf)
        return bodies

    # Create the coroutines
    send_crt = send_requests()
    recv_crt = read_responses()

    # Register the coroutines with the event loop
    send_future = AioFuture(send_crt, loop=loop)
    recv_future = AioFuture(recv_crt, loop=loop)

    # Run the event loop until the receive coroutine is done (which
    # implies that all the requests must have been sent as well):
    loop.run_until_complete(recv_future)

    # Get the result returned by the coroutine
    bodies = recv_future.result()

# end-example


examples/extract_links.py
#!/usr/bin/env python3
'''
Extract all links from a URL.
'''

import sys
import os.path
from io import TextIOWrapper
from html.parser import HTMLParser
from urllib.parse import urlsplit, urljoin, urlunsplit
import re
import ssl

# We are running from the dugong source directory, append it to module path so
# that we can fallback on it if dugong hasn't been installed yet.
if __name__ == '__main__':
    basedir = os.path.abspath(os.path.join(os.path.dirname(sys.argv[0]), '..'))
else:
    basedir = os.path.abspath(os.path.join(os.path.dirname(__file__), '..'))
if (os.path.exists(os.path.join(basedir, 'setup.py')) and
    os.path.exists(os.path.join(basedir, 'dugong', '__init__.py'))):
    sys.path.append(basedir)

from dugong import HTTPConnection

# When running from HG repo, enable all warnings
if os.path.exists(os.path.join(basedir, '.hg')):
    import warnings
    warnings.simplefilter('error')

class LinkExtractor(HTMLParser):
    def __init__(self):
        if sys.version_info < (3,4):
            # Python 3.3 doesn't know about convert_charrefs
            super().__init__()
        else:
            super().__init__(convert_charrefs=True)
        self.links = []

    def handle_starttag(self, tag, attrs):
        if tag != 'a':
            return

        for (name, val) in attrs:
            if name == 'href':
                self.links.append(val)
            break

def main():
    if len(sys.argv) != 2:
        raise SystemExit('Usage: %s <url>' % sys.argv[0])
    url = sys.argv[1]
    url_els = urlsplit(url)

    if url_els.scheme == 'https':
        ssl_context = ssl.SSLContext(ssl.PROTOCOL_SSLv23)
        ssl_context.options |= ssl.OP_NO_SSLv2
        ssl_context.verify_mode = ssl.CERT_REQUIRED
        ssl_context.set_default_verify_paths()
    else:
        ssl_context = None

    with HTTPConnection(url_els.hostname, port=url_els.port,
                          ssl_context=ssl_context) as conn:
        path = urlunsplit(('', '') + url_els[2:4] + ('',)) or '/'
        conn.send_request('GET', path)
        resp = conn.read_response()
        if resp.status != 200:
            raise SystemExit('%d %s' % (resp.status, resp.reason))

        # Determine if we're reading text or binary data, and (in case of text),
        # what character set is being used.
        if 'Content-Type' not in resp.headers:
            type_ = 'application/octet-stream'
        else:
            type_ = resp.headers['Content-Type']

        hit = re.match(r'text/x?html(?:; charset=(.+))?$', type_)
        if not hit:
            raise SystemExit('Server did not send html but %s' % type_)

        if hit.group(1):
            charset = hit.group(1)
        else:
            charset = 'latin1'

        html_stream = TextIOWrapper(conn, encoding=charset)
        parser = LinkExtractor()

        while True:
            buf = html_stream.read(16*1024)
            if not buf:
                break
            parser.feed(buf)

    for link in parser.links:
        print(urljoin(url, link))

if __name__ == '__main__':
    main()
